chore: alternative  algorithm for get max rate for streams

olelod opened the pull request:
The get max standard rate algorithm for a multiple streams and pressures compressor train does not work fully as intended. Changing to a brute force algorithm, instead of trying to be clever in searching along max speed curve and stone wall etc (old code over 400 lines, this one around 60-70).

This algorithm will of course be more computationally demanding.

## Have you remembered and considered?

- [ ] I have remembered to update documentation
- [ ] I have remembered to update manual changelog (`docs/drafts/next.draft.md`)
- [ ] I have remembered to update migration guide (`docs/docs/migration_guides/`)
- [ ] I have committed with `BREAKING:` in footer or `!` in header, if breaking
- [ ] I have added tests (if not, comment why)
- [ ] I have used conventional commits syntax (if you squash, make sure that conventional commit is used)
- [ ] I have included the Jira issue ID somewhere in the commit body (`ECALC-XXXX`)

## Why is this pull request needed?

This pull request is needed because of....

## What does this pull request change?

Write summary of what this pull request changes if needed.

## Issues related to this change:

Refs.
ECALC-416

jsolaas commented on src/libecalc/core/models/compressor/train/variable_speed_compressor_train_common_shaft_multiple_streams_and_pressures.py:
```diff
@@ -405,6 +407,70 @@ def get_max_standard_rate_per_stream(
+        # if it is not an ingoing stream --> currently no calculations done
+        # Fixme: what should be returned? 0.0, NaN or something else?
+        if not self.streams[stream_to_maximize].is_inlet_stream:
+            return 0.0
```
raise error? If that happens a developer made a mistake?

olelod replied:
It's not an error. This is if we take something out (e.g. for export) between two compressors. I guess in reality we will be limited by what has entered the compressor train earlier - we could in theory take everything out - so it could be the sum of the rates for all the streams that has entered/left the compressor train at earlier stages.

jsolaas commented on src/libecalc/core/models/compressor/train/variable_speed_compressor_train_common_shaft_multiple_streams_and_pressures.py:
```diff
@@ -405,6 +407,70 @@ def get_max_standard_rate_per_stream(
+                )
+        else:
+            max_rate_is_larger_than = std_rates_std_m3_per_day_per_stream[stream_to_maximize]
+        while train_result.is_valid:
+            max_rate_is_larger_than = train_result.stage_results[
+                stream_to_maximize_connected_to_stage_no
+            ].standard_rate_asv_corrected_sm3_per_day
+            std_rates_std_m3_per_day_per_stream[stream_to_maximize] = max_rate_is_larger_than * 2
+            train_result = self.calculate_shaft_speed_given_rate_ps_pd(
+                std_rates_std_m3_per_day_per_stream=std_rates_std_m3_per_day_per_stream,
+                suction_pressure=suction_pressure,
+                target_discharge_pressure=target_discharge_pressure,
+            )
```
Is this to find a rough max for the maximize function?

olelod replied:
We know what the rate should be larger than, so we double it until we reach something that is too large. Then we have an interval that we know is_valid will switch from true to false in - and we start looking for the point where it switches.

TeeeJay replied:
What is the convergence criteria? EPSILON?

olelod replied:
The interval is halved and we continue in the upper or lower half depending on if the evaluation in the middle is true (upper half) or false (lower half). The evaluation stops when (highest value -lowest value)/highest value is less than 1e-5 or after 50 iterations.

jsolaas commented on src/libecalc/core/models/compressor/train/variable_speed_compressor_train_common_shaft_multiple_streams_and_pressures.py:
```diff
@@ -405,6 +407,70 @@ def get_max_standard_rate_per_stream(
+                suction_pressure=suction_pressure,
+                target_discharge_pressure=target_discharge_pressure,
+            )
+        return maximize_x_given_boolean_condition_function(
+            x_min=float(max_rate_is_larger_than),
+            x_max=float(max_rate_is_larger_than * 2),
+            bool_func=lambda x: _calculate_train_result(std_rate_for_stream=x).is_valid,
+        )
```
And this find a more accurate max rate?

olelod replied:
And here we find the point where it switches trom true to false

TeeeJay replied:
Among the "points we have" or as soon as we are within EPSILON of "where it switches"?

jsolaas reviewed (approved):
My vote is to remove the previous implementation

olelod commented:
> My vote is to remove the previous implementation

Done

TeeeJay commented:
> The get max standard rate algorithm for a multiple streams and pressures compressor train does not work fully as intended. Adding a brute force algorithm, instead of trying to be clever in searching along max speed curve and stone wall etc (old code over 400 lines, this one around 60-70).
> 
> This algorithm will of course be morw computationally demanding.
> 
> Have currently not removed the old code (its currently there but not used), just added a new one.
> ## Have you remembered and considered?
> 
>     * [ ]  I have remembered to update documentation
> 
>     * [ ]  I have remembered to update manual changelog (`docs/drafts/next.draft.md`)
> 
>     * [ ]  I have remembered to update migration guide (`docs/docs/migration_guides/`)
> 
>     * [ ]  I have committed with `BREAKING:` in footer or `!` in header, if breaking
> 
>     * [ ]  I have added tests (if not, comment why)
> 
>     * [ ]  I have used conventional commits syntax (if you squash, make sure that conventional commit is used)
> 
>     * [ ]  I have included the Jira issue ID somewhere in the commit body (`ECALC-XXXX`)
> 
> 
> ## Why is this pull request needed?
> 
> This pull request is needed because of....
> ## What does this pull request change?
> 
> Write summary of what this pull request changes if needed.
> ## Issues related to this change:
> 
> Refs. ECALC-416

What are the known limitations of the previous implementation? As I understood, it is either very difficult or complicated, or not even possible to make the previous version work as intended?

olelod commented:
> > The get max standard rate algorithm for a multiple streams and pressures compressor train does not work fully as intended. Adding a brute force algorithm, instead of trying to be clever in searching along max speed curve and stone wall etc (old code over 400 lines, this one around 60-70).
> > This algorithm will of course be morw computationally demanding.
> > Have currently not removed the old code (its currently there but not used), just added a new one.
> > ## Have you remembered and considered?
> > ```
> > * [ ]  I have remembered to update documentation
> > 
> > * [ ]  I have remembered to update manual changelog (`docs/drafts/next.draft.md`)
> > 
> > * [ ]  I have remembered to update migration guide (`docs/docs/migration_guides/`)
> > 
> > * [ ]  I have committed with `BREAKING:` in footer or `!` in header, if breaking
> > 
> > * [ ]  I have added tests (if not, comment why)
> > 
> > * [ ]  I have used conventional commits syntax (if you squash, make sure that conventional commit is used)
> > 
> > * [ ]  I have included the Jira issue ID somewhere in the commit body (`ECALC-XXXX`)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ## Why is this pull request needed?
> > This pull request is needed because of....
> > ## What does this pull request change?
> > Write summary of what this pull request changes if needed.
> > ## Issues related to this change:
> > Refs. ECALC-416
> 
> What are the known limitations of the previous implementation? As I understood, it is either very difficult or complicated, or not even possible to make the previous version work as intended?

It's not impossible, I would think, but complicated. You need to keep track of the previous streams and subtract them when finding the maximum rate for a stream in the middle of the compressor train. And it will be dependent of pressure, temperature and rotational speed in the train. It just did not work for all cases - it seemed, and then we choose to go to a simpler implementation.
